### PR TITLE
Move tolerations to PodSpec for ip-masq-agent.yaml.

### PR DESCRIPTION
**Pull request by gyliu513**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-unit test this

**Comment by gmarek** [comment]
I'm curious about this PR, and actually I'm more curious about one that added this. I don't think we taint master machines, so I don't think those tolerations are needed at all. @dnardo - how did you ended up with them?

**Comment by gyliu513** [comment]
@gmarek I think it is a valid case to make the master have some taint, the reason is that sometimes we only want master to run some system services and for those services, we need to add tolerations for it to make sure only system services can run on master while other non-system services will be scheduled on to those nodes.

**Comment by gmarek** [comment]
@gyliu513 - we currently use Unschedulable bit for preventing scheduling on master. Let's not add Tolerations before we know what kind of Taints there are on master.
@kubernetes/sig-cluster-lifecycle-misc - are there any plans on tainting master?

**Comment by gyliu513** [comment]
@gmarek If using Unschedulable bit, then we will not able to deploy some system services on master node, this is my major concern for this. How to handle the case if I want to deploy some system services as deployment to master node?

**Comment by gmarek** [comment]
@gyliu513 - you need to use manifest files for this. Plus daemons ignore Unschedulable bit.

**Comment by gyliu513** [comment]
Thanks @gmarek Yes, that is option, but sometimes the static pod (manifest you mentioned) or daemonset may not fulfill my requirement for the following reasons:
1) With static pod, I cannot use `kubectl edit` to do on-line edit, but have to logon to each node and edit the manifest manually, this is not convenient for the cluster admin.
2) Sometimes, I may want to group all of the master nodes to different groups and each group may have different system services.

So seems deployment is a good choice as it support both on-line edit and selective deployment. Comments?

**Comment by gmarek** [comment]
It's not that I don't want to taint master nodes. I'm just saying that we're not doing it now, so we should add Tolerations before we know if/how we're going to do it. @kubernetes/sig-cluster-lifecycle-feature-requests is a place to ask (I guess).

**Comment by gyliu513** [comment]
Got it, thanks @gmarek, but for this daemonset, I think that it is not limited to master node but also worker node, right?

**Comment by gmarek** [comment]
Yest, but one of those Tolerations is for master only, and second we also don't put anywhere AFAICT.

**Comment by gyliu513** [comment]
ok, then I think that we can remove the master taint, but for the `CriticalAddonsOnly`, I think that we still need to keep it for rescheduler. Please refer to https://kubernetes.io/docs/tasks/administer-cluster/guaranteed-scheduling-critical-addon-pods/ for detail. 

This is the info from above link:
>To avoid situation when another pod is scheduled into the space prepared for the critical add-on, the chosen node gets a temporary taint “CriticalAddonsOnly” before the eviction(s) (see more details). Each critical add-on has to tolerate it, while the other pods shouldn’t tolerate the taint. The taint is removed once the add-on is successfully scheduled.

Comments @gmarek

**Comment by gmarek** [comment]
OK, thanks for the pointer.

**Comment by thockin** [comment]
https://github.com/kubernetes/kubernetes/pull/35211

I guess we can remove the master part

**Comment by dnardo** [comment]
Yes, sorry this is my fault, I had seen other DaemonSet's doing this and though it was a "thing" turns out it's not

File: cluster/addons/ip-masq-agent/ip-masq-agent.yaml [hunk 1]
```diff
@@ -36,3 +33,9 @@ spec:
               # The daemon looks for its config in a YAML file at /etc/config/ip-masq-agent
               - key: config
                 path: ip-masq-agent
+      tolerations:
+        - key: "CriticalAddonsOnly"
+          operator: "Exists"
+        - key: "dedicated"
+          value: "master"
+          effect: "NoSchedule"
```
Comment (dnardo) [inline cluster/addons/ip-masq-agent/ip-masq-agent.yaml]: Lets remove the master part as suggested by Marek, but lets leave the CriticalAddonsOnly please.

**Review by dnardo — CHANGES_REQUESTED** [review]

**Comment by luxas** [comment]
FYI kubeadm, kops, kargo and bootkube AFAIK uses the `node-role.kubernetes.io/master="":NoSchedule` taint. Might be worth keeping that toleration in that perspective.

In v1.5 while in alpha, we initially used `dedicated=master:NoSchedule` but switched because that taint should AFAIK be left to the user.

File: cluster/addons/ip-masq-agent/ip-masq-agent.yaml (hunk 1 above)
Comment (gyliu513) [inline cluster/addons/ip-masq-agent/ip-masq-agent.yaml]: Done

**Comment by gyliu513** [comment]
Thanks @luxas for pointing this out!

@dnardo @gmarek @thockin any comments for this? I think we may need to support those master taints, comments? 

```yaml
tolerations:
- key: "CriticalAddonsOnly"
  operator: "Exists"
- key: "dedicated"
  value: "master"
  effect: "NoSchedule"
- key: "node-role.kubernetes.io/master"
  effect: "NoSchedule"
```

**Comment by dnardo** [comment]
@gyliu513 Looking through the PR Tim linked, It sounds like those tools should not be relying on this so it seems like the tools themselves should find a better way to handle this. Either that or #35211 should get resolved.

I guess for now we should remove it

**Comment by gyliu513** [comment]
Done, thanks @dnardo

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this
@k8s-bot pull-kubernetes-unit test this

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-unit test this
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by gyliu513** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by gyliu513** [comment]
@thockin can you please help mark this as 1.7 milestone? Otherwise, it will not be put into submit queue.

**Comment by thockin** [comment]
@dchen1107 This is a bugfix, functionally no-op, but correct yaml syntax.

**Pull request merged**